frontend/src/pages/queryBuilder/GraphEditor.jsx

```diff
@@ -0,0 +1,187 @@
+import React, {
+  useContext, useReducer, useEffect,
+} from 'react';
+import Popover from '@material-ui/core/Popover';
+import Button from '@material-ui/core/Button';
+import Paper from '@material-ui/core/Paper';
+
+import QueryBuilderContext from '~/context/queryBuilder';
+import nodeUtils from '~/utils/d3/nodes';
+
+import QueryGraph from './QueryGraph';
+import NodeSelector from './textEditorRow/NodeSelector';
+import PredicateSelector from './textEditorRow/PredicateSelector';
+
+import './graphEditor.css';
+
+const width = 600;
+const height = 400;
+
+function reducer(state, action) {
+  switch (action.type) {
+    case 'startConnection': {
+      const { anchor } = action.payload;
+      state.creatingConnection = true;
+      state.popoverId = '';
+      state.popoverAnchor = anchor;
+      state.popoverType = 'newEdge';
+      break;
+    }
+    case 'connectTerm': {
+      const { id } = action.payload;
+      state.chosenTerms = [...state.chosenTerms, id];
+      break;
+    }
+    case 'connectionMade': {
+      state.creatingConnection = false;
+      state.chosenTerms = [];
+      break;
+    }
+    case 'click': {
+      const { id } = action.payload;
+      state.clickedId = id;
+      break;
+    }
+    case 'openEditor': {
+      const { id, type, anchor } = action.payload;
+      state.popoverId = id;
+      state.popoverType = type;
+      state.popoverAnchor = anchor;
+      break;
+    }
+    case 'closeEditor': {
+      state.popoverId = '';
+      state.popoverType = '';
+      state.popoverAnchor = null;
+      break;
+    }
+    default: {
+      return state;
```

Comment:
That is actually probably desired. Returning the same reference to state in the default case here will not cause a rerender. If we were to make a copy and return, it would trigger a rerender of the same exact state.